File: VBotMauiApp/Platforms/Android/OngoingCallNotification.cs
using Android.App;
using Android.Content;
using Android.OS;
using AndroidX.Core.App;

namespace VBotMauiApp.Platforms.Android;

public static class OngoingCallNotification
{
    private const string ChannelId = "vbot_ongoing_call_channel";
    private const int NotificationId = 9999;

    public static void Show(Context context, string callerName)
    {
        var appContext = context.ApplicationContext ?? context;
        var notificationManager = appContext.GetSystemService(Context.NotificationService) as NotificationManager;
        if (notificationManager == null) return;

        if (Build.VERSION.SdkInt >= BuildVersionCodes.O)
        {
            var channel = new NotificationChannel(
                ChannelId,
                "Cuộc gọi đang diễn ra",
                NotificationImportance.Low
            )
            {
                Description = "Hiển thị thông báo khi đang trong cuộc gọi VBot"
            };
            channel.SetShowBadge(false);
            channel.SetSound(null, null);
            channel.EnableVibration(false);
            notificationManager.CreateNotificationChannel(channel);
        }

        var openAppIntent = new Intent(appContext, typeof(MainActivity));
        openAppIntent.AddFlags(ActivityFlags.SingleTop | ActivityFlags.NewTask);
        var contentPendingIntent = PendingIntent.GetActivity(
            appContext, 0, openAppIntent,
            PendingIntentFlags.Immutable | PendingIntentFlags.UpdateCurrent
        );

        var hangupIntent = new Intent(appContext, typeof(HangupReceiver));
        var hangupPendingIntent = PendingIntent.GetBroadcast(
            appContext, 1, hangupIntent,
            PendingIntentFlags.Immutable
        );

        var builder = new NotificationCompat.Builder(appContext, ChannelId)
            .SetSmallIcon(global::Android.Resource.Drawable.IcMenuCall)
            .SetContentTitle("Đang trong cuộc gọi")
            .SetContentText(string.IsNullOrEmpty(callerName) ? "VBot Phone" : callerName)
            .SetOngoing(true)
            .SetCategory(NotificationCompat.CategoryCall)
            .SetContentIntent(contentPendingIntent)
            .AddAction(global::Android.Resource.Drawable.IcMenuCloseClearCancel, "Tắt máy", hangupPendingIntent)
            .SetPriority(NotificationCompat.PriorityLow)
            .SetVisibility(NotificationCompat.VisibilityPublic);

        notificationManager.Notify(NotificationId, builder.Build());
    }

    public static void Cancel(Context context)
    {
        var appContext = context.ApplicationContext ?? context;
        var notificationManager = appContext.GetSystemService(Context.NotificationService) as NotificationManager;
        notificationManager?.Cancel(NotificationId);
    }
}

File: android/vbot-android-wrapper/src/main/kotlin/com/vpmedia/vbot/wrapper/OngoingCallNotification.kt
package com.vpmedia.vbot.wrapper

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import androidx.annotation.Keep
import androidx.core.app.NotificationCompat

@Keep
object OngoingCallNotification {
    private const val CHANNEL_ID = "vbot_ongoing_call"
    const val NOTIFICATION_ID = 9999

    fun show(context: Context, callerName: String, targetActivityClass: Class<*>?) {
        val appContext = context.applicationContext
        val notificationManager =
            appContext.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                CHANNEL_ID,
                "Cuộc gọi đang diễn ra",
                NotificationManager.IMPORTANCE_LOW
            ).apply {
                setShowBadge(false)
                setSound(null, null)
                enableVibration(false)
            }
            notificationManager.createNotificationChannel(channel)
        }

        val launchIntent = if (targetActivityClass != null) {
            Intent(appContext, targetActivityClass).apply {
                addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP or Intent.FLAG_ACTIVITY_NEW_TASK)
            }
        } else {
            appContext.packageManager.getLaunchIntentForPackage(appContext.packageName)?.apply {
                addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP or Intent.FLAG_ACTIVITY_NEW_TASK)
            } ?: Intent()
        }

        val contentIntent = PendingIntent.getActivity(
            appContext,
            0,
            launchIntent,
            PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
        )

        val hangupIntent = PendingIntent.getBroadcast(
            appContext,
            1,
            Intent(appContext, HangupReceiver::class.java),
            PendingIntent.FLAG_IMMUTABLE
        )

        val notification = NotificationCompat.Builder(appContext, CHANNEL_ID)
            .setSmallIcon(android.R.drawable.ic_menu_call)
            .setContentTitle("Đang trong cuộc gọi")
            .setContentText(callerName)
            .setOngoing(true)
            .setCategory(NotificationCompat.CATEGORY_CALL)
            .setContentIntent(contentIntent)
            .addAction(
                android.R.drawable.ic_menu_close_clear_cancel,
                "Tắt máy",
                hangupIntent
            )
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
            .build()

        notificationManager.notify(NOTIFICATION_ID, notification)
    }

    fun cancel(context: Context) {
        val notificationManager = context.applicationContext
            .getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        notificationManager.cancel(NOTIFICATION_ID)
    }
}
